### PR TITLE
Fix linting for `modules.json` to add support for only installing subworkflows from a repository

### LINKED ISSUE #4286: Linting for `modules.json` fails when no modules have been installed from a remote repository
### Description of the bug

`modules_json()` checks for a `modules` entry:https://github.com/nf-core/tools/blob/b6c5737fde996c824b87d19d1c51a29b4810db1d/nf_core/pipelines/lint/modules_json.py#L29-L34

 However, it's possible to have only installed subworkflows and not modules from a repository, for example: 

```
    "repos": {
        "https://github.com/genomic-medicine-sweden/nf-core-modules.git": {
            "subworkflows": {
                "genomic-medicine-sweden": {
                    "vcf_annotate_score_genmod": {
                        "branch": "main",
                        "git_sha": "37fef772cb8e2480f301c201c86f36d8e1c94349",
                        "installed_by": ["subworkflows"]
                    }
                }
            }
        },
        "https://github.com/nf-core/modules.git": {
            "modules": {
                "nf-core": {
...
```

In that case `nf-core pipelines lint` fails.

### Command used and terminal output

```console

```

### System information

_No response_

### PR DESCRIPTION
**Pull request by beatrizsavinhas**
Closes https://github.com/nf-core/tools/issues/4286.

Improve linting for `modules.json`:

* Add support for repositories from where only `subworkflows` have been installed since this was currently not allowed.
* Add explicit linting error message for when there are either no `modules` or `subworkflows` installed for a repository.
* Add explicit linting error messages for repository links that don't start with "http" - this was an existing check that only returned a general error message.

## PR checklist

- [x] This comment contains a description of changes (with reason)
- [x] `CHANGELOG.md` is updated
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] Documentation in `docs` is updated

### CODE REVIEW

File: nf_core/pipelines/lint/modules_json.py [hunk 1]
```diff
@@ -25,32 +25,32 @@ def modules_json(self) -> dict[str, list[str]]:
...
-                            f"Entry for `{Path(modules_dir, install_dir, module)}` is missing branch information."
-                        )
-                    if module_entry.get("git_sha") is None:
-                        failed.append(
-                            f"Entry for `{Path(modules_dir, install_dir, module)}` is missing version information."
-                        )
-        if all_modules_passed:
-            passed.append("Only installed modules found in `modules.json`")
+            if not repo.startswith("http"):
+                failed.append(f'Repository link for {repo} doesn\'t start with "http" in `modules.json`.')
```
Comment (beatrizsavinhas) [inline nf_core/pipelines/lint/modules_json.py:29]: This change is arguably out of scope for the PR however I thought it simplified the logic.

**Comment by mashehu** [comment]
@nf-core-bot fix linting

**Review by mashehu — COMMENTED** [review]
Thanks for the contribution, looks good!
Could you please also add a test case for a modules_json with only subworkflow entries (and one completely empty ones) to `tests/modules/test_modules_json.py`?

**Comment by beatrizsavinhas** [comment]
> Thanks for the contribution, looks good! Could you please also add a test case for a modules_json with only subworkflow entries (and one completely empty ones) to `tests/modules/test_modules_json.py`?

Very fair to request tests! We were trying to quickly patch the bug to move on with development. But I'll spend some extra time adding tests as I do agree it would make the codebase better.

**Comment by mashehu** [comment]
sorry the last commit by me was meant for a different branch...

File: tests/pipelines/lint/test_modules_json.py [hunk 2]
```diff
@@ -1,10 +1,45 @@
+import json
+
+from nf_core.modules.modules_json import (
+    ModulesJson,
+    ModulesJsonType,
```
Comment (beatrizsavinhas) [inline tests/pipelines/lint/test_modules_json.py:5]: There's no equivalent for `NFCoreYamlConfig` from `nf_core.utils` the modules json file so I used these classes directly.

File: tests/pipelines/lint/test_modules_json.py [hunk 3]
```diff
@@ -1,10 +1,45 @@
...
+        for repo in new_modules_json["repos"]:
+            new_modules_json["repos"][repo] = {"subworkflows": new_modules_json["repos"][repo]["subworkflows"]}
+
+        self._update_modules_json(new_modules_json)
+
+        # Check linting results
+        results = self.lint_obj.modules_json()
+        assert len(results.get("warned", [])) == 0
+        assert len(results.get("failed", [])) == 0
+        assert len(results.get("passed", [])) == 0
```
Comment (beatrizsavinhas) [inline tests/pipelines/lint/test_modules_json.py:74]: With the current set up there is no subworkflow linting so no `passed` is returned. Should we open an issue to add subworkflow linting in the future?

File: tests/pipelines/lint/test_modules_json.py [hunk 4]
```diff
@@ -1,10 +1,45 @@
...
 from ..test_lint import TestLint
 
 
 class TestLintModulesJson(TestLint):
+    def setUp(self) -> None:
+        super().setUp()
+        self._modules_json_object: ModulesJson = ModulesJson(self.pipeline_dir)
+        self._modules_json_object.load()
+        self.modules_json_path = self._modules_json_object.modules_json_path
+        self.modules_json: ModulesJsonType | None = self._modules_json_object.modules_json
```
Comment (beatrizsavinhas) [inline tests/pipelines/lint/test_modules_json.py]: This attribute can't actually be `None` due to `.load()` on line 15, but mypy throws an error since it is inherited from `ModulesJson.modules_json` which has `ModulesJsonType | None` typehint.

File: tests/pipelines/lint/test_modules_json.py (hunk 3 above)
Comment (mashehu) [reply tests/pipelines/lint/test_modules_json.py:74]: yes, please do that.

File: tests/pipelines/lint/test_modules_json.py (hunk 4 above)
Comment (mashehu) [reply tests/pipelines/lint/test_modules_json.py]: you can just throw in an `assert` to make mypy happy

File: tests/pipelines/lint/test_modules_json.py (hunk 3 above)
Comment (beatrizsavinhas) [reply tests/pipelines/lint/test_modules_json.py:74]: Done here https://github.com/nf-core/tools/issues/4293!

File: tests/pipelines/lint/test_modules_json.py (hunk 4 above)
Comment (beatrizsavinhas) [reply tests/pipelines/lint/test_modules_json.py]: Done now! See line 16.
Thanks for the tip!

File: tests/pipelines/lint/test_modules_json.py [hunk 5]
```diff
@@ -1,10 +1,74 @@
...
+        new_modules_json = self.modules_json
+        first_repo = next(iter(new_modules_json["repos"]))
+        new_modules_json["repos"]["incorrect_repo_link"] = new_modules_json["repos"].pop(first_repo)
+
+        self._update_modules_json(new_modules_json)
+
+        # Check linting results
+        results = self.lint_obj.modules_json()
+        assert len(results.get("warned", [])) == 0
+        assert len(results.get("failed", [])) == 1
```
Comment (mashehu) [inline tests/pipelines/lint/test_modules_json.py:45]: to be extra precise, we sometimes also check the failed test name: https://github.com/nf-core/tools/blob/dev/tests/modules/lint/test_meta_yml.py#L71

Comment (mashehu) [reply tests/pipelines/lint/test_modules_json.py:45]: but the current asserts are already good

**Review by mashehu — APPROVED** [review]
Looking good now. just a small comment about additional possible asserts, but good to go as it is 🙂

File: tests/pipelines/lint/test_modules_json.py (hunk 5 above)
Comment (beatrizsavinhas) [reply tests/pipelines/lint/test_modules_json.py:45]: Yeah, I thought it could be nice to check the errors more thoroughly but in this case they are just strings so the test would be very flimsy in case the error strings are updated in the future.
In the example you gave, `module_lint` is a class in itself whereas here `results` is just a dict containing strings. I think implementing checks like what is done for `meta_yml` would require more complex changes that would be outside of the scope of this PR. 🤔
